### PR TITLE
feat: build column lineage using sqlglot 

### PR DESCRIPTION
**Pull request by eyelesbarrow**
This extracts and build column level lineage data from  a SAP Hana db using sqlglot.  
The column lineage is built from  sqlglot nodes, which contains upstream and downstream data, before being transformed into workunits to be sent to datahub. 
At this point, code is still WIP, with testing and experimentation for edge cases still ongoing.

### CODE REVIEW

File: LICENSE
Comment (LucasRoesler) [inline LICENSE:1]: I don't understand why this file is changing

File: datahub_sap_hana/hana_recipe.yaml
Comment (LucasRoesler) [inline datahub_sap_hana/hana_recipe.yaml:1]: i think there are too many recipes being included in the project, it seems reasonable that we have exactly _two_ example recipes in the project root, one for console output and one for the datahub server which is configured by env variables. You shouldn't need to add any more after that.

File: pyproject.toml [hunk 1]
```diff
@@ -30,7 +31,9 @@ build-backend = "poetry.core.masonry.api"
 [tool.pytest.ini_options]
 testpaths = [
     "tests/unit",
-    "tests/integration"
+    "tests/integration",
+    "integration: marks test as integration tests (select with '-m integration')"
+
```
Comment (LucasRoesler) [inline pyproject.toml:36]: This looks like a bad conflict resolution, you can delete it
```suggestion
    "tests/integration"
```

File: tests/integration/data/hana_mces_golden_lineage_disabled.json
Comment (LucasRoesler) [inline tests/integration/data/hana_mces_golden_lineage_disabled.json:1]: seems odd that the timestamps change, i thought the tests used a frozen time value

File: tests/test_helpers/click_helpers.py [hunk 2]
```diff
@@ -1,9 +1,3 @@
-# This is copied and slightly modified from the acryl-datahub project
```
Comment (LucasRoesler) [inline tests/test_helpers/click_helpers.py:1]: why are you removing the license headers? these should not be removed from the files that have them

File: datahub_sap_hana/col_lineage/sqlglot_test.py
Comment (LucasRoesler) [inline datahub_sap_hana/col_lineage/sqlglot_test.py:1]: i don't think this file is needed and should be replaced with a real test in the `tests` folder instead

**Review by LucasRoesler — CHANGES_REQUESTED** [review]
just the first round of suggestions that are more about project logistics and file structure. I will read through the new ingestion logic now

File: datahub_sap_hana/ingestion.py [hunk 3]
```diff
@@ -178,3 +200,115 @@ def _get_view_lineage_workunits(self) -> Iterable[MetadataWorkUnit]:
...
+
+        with engine.connect() as conn:
+            inspector = inspect(conn)
+            schema = inspector.get_schema_names()  # returns a list
+            views = inspector.get_view_names(
+                schema[0])  # returns a list
+
+            for view in views:
+                definition = inspector.get_view_definition(view, schema[0])
+                yield (view, definition)
```
Comment (LucasRoesler) [inline datahub_sap_hana/ingestion.py:217]: Is `view` the name? You can make the code a little easier to read by changing the variable name here
```suggestion
            for view_name in views:
                definition = inspector.get_view_definition(view_name, schema[0])
                yield (view_name, definition)
```

File: datahub_sap_hana/ingestion.py [hunk 4]
```diff
@@ -178,3 +200,115 @@ def _get_view_lineage_workunits(self) -> Iterable[MetadataWorkUnit]:
...
+    def get_column_lineage_view_definitions(self) -> Iterable[Tuple[str, str]]:
+
+        url = self.config.get_sql_alchemy_url()
+        engine = create_engine(url, **self.config.options)
+
+        with engine.connect() as conn:
+            inspector = inspect(conn)
+            schema = inspector.get_schema_names()  # returns a list
+            views = inspector.get_view_names(
+                schema[0])  # returns a list
```
Comment (LucasRoesler) [inline datahub_sap_hana/ingestion.py:213]: this indenting looks wrong, make sure to run `task lint` and `task fix` before opening the PRs

Comment (LucasRoesler) [inline datahub_sap_hana/ingestion.py:213]: why only `schema[0]` couldn't there be multiple schemas? also it should use the schema name pattern filter here to decide which schemas to process and which to skip.

File: datahub_sap_hana/ingestion.py [hunk 5]
```diff
@@ -178,3 +200,115 @@ def _get_view_lineage_workunits(self) -> Iterable[MetadataWorkUnit]:
...
+                yield (view, definition)
+                # print(view, definition)
+
+    def get_lineage_for_column(self, view_name: str, view_sql: str) -> Iterable[Tuple[str, Any]]:
+
+        cols = parse_one(view_sql).named_selects
+        for col_name in cols:
+            yield (view_name, lineage(col_name.lower(), view_sql.lower()))
+
+    def parse_column_name(self, column_name: str):
```
Comment (LucasRoesler) [inline datahub_sap_hana/ingestion.py:226]: this function definitely requires a doc string explaining _why_ we do this parsing.

File: datahub_sap_hana/ingestion.py [hunk 6]
```diff
@@ -178,3 +200,115 @@ def _get_view_lineage_workunits(self) -> Iterable[MetadataWorkUnit]:
                 wu = item.as_workunit()
                 self.report.report_workunit(wu)
                 yield wu
+
+    def get_column_lineage_view_definitions(self) -> Iterable[Tuple[str, str]]:
+
+        url = self.config.get_sql_alchemy_url()
+        engine = create_engine(url, **self.config.options)
+
+        with engine.connect() as conn:
```
Comment (LucasRoesler) [inline datahub_sap_hana/ingestion.py:209]: I think it makes sense that your functions accept a `conn` connection instead of building one. Normally you don't want to create new connections frequently, instead you make a single connection pool `conn` and pass that around. I would do that in the `get_workunits`, build the `conn` and then pass it to the `_get_view_lineage_workunits` and any other functions that need it.

File: datahub_sap_hana/ingestion.py [hunk 7]
```diff
@@ -178,3 +200,115 @@ def _get_view_lineage_workunits(self) -> Iterable[MetadataWorkUnit]:
...
+            schema = inspector.get_schema_names()  # returns a list
+            views = inspector.get_view_names(
+                schema[0])  # returns a list
+
+            for view in views:
+                definition = inspector.get_view_definition(view, schema[0])
+                yield (view, definition)
+                # print(view, definition)
+
+    def get_lineage_for_column(self, view_name: str, view_sql: str) -> Iterable[Tuple[str, Any]]:
```
Comment (LucasRoesler) [inline datahub_sap_hana/ingestion.py:220]: I am not sure this function is named correctly. I think `_get_column_lineage_for_view` make more sense given the input arguments. 

Also, I think this is one where we want to include a doc string about what is happening.

File: datahub_sap_hana/ingestion.py [hunk 8]
```diff
@@ -178,3 +200,115 @@ def _get_view_lineage_workunits(self) -> Iterable[MetadataWorkUnit]:
...
+
+        cols = parse_one(view_sql).named_selects
+        for col_name in cols:
+            yield (view_name, lineage(col_name.lower(), view_sql.lower()))
```
Comment (LucasRoesler) [inline datahub_sap_hana/ingestion.py:224]: i think it might be easier to not use `yield` here and instead build an actual list of lineages

```suggestion
        view_sql = view_sql.lower()
        lineages = [ 
           lineage(col_name.lower), view_sql)
           for col_name in cols
        ]
        return view_name, lineages
```

It might even make sense to change how much lineage is returned, you could do something like this because we _only_ really care about the first downstream list, we don't need the whole lineage node do we?

```
        view_sql = view_sql.lower()
        lineages = [ 
           (col_name.lower, lineage(col_name.lower), view_sql).downstream)
           for col_name in cols
        ]
        return view_name, lineages
```

this gives you return type that looks like `Tuple[str, List[str, List[Node]]]`

meaning it looks like 

```
(
    view_name,
    [
      (
          column_one, 
          [node1, node2]
      ),
      (
         column_two,
         [another_node]
      )
    ]
)
```

The benefit of this output is that the column lineage stays grouped by the view and you only get the data that you need, the first step in the lineage for each column.

File: datahub_sap_hana/ingestion.py [hunk 9]
```diff
@@ -178,3 +200,115 @@ def _get_view_lineage_workunits(self) -> Iterable[MetadataWorkUnit]:
...
+
+    def parse_column_name(self, column_name: str):
+        parsed_name = column_name.split(".")
+        return parsed_name[1]
+
+    def get_column_view_lineage_elements(self) -> Iterable[Tuple[Field, List[UpstreamField]]]:
```
Comment (LucasRoesler) [inline datahub_sap_hana/ingestion.py:230]: for functions that return rather nested types like this, i recommend writing a doc string and include a description of what the return type actually mean, sometimes even an pseudo-code example.

File: datahub_sap_hana/ingestion.py [hunk 10]
```diff
@@ -178,3 +200,115 @@ def _get_view_lineage_workunits(self) -> Iterable[MetadataWorkUnit]:
...
+if __name__ == "__main__":
+
+    hana_config = HanaConfig(username="HOTEL",
+                             password="Localdev1",
+                             host_port="localhost:39041",
+                             database="HXE",
+                             schema_pattern=AllowDenyPattern(allow=["*"]),
+                             include_view_lineage=True,
+                             )  # type: ignore
+    hana_source = HanaSource(hana_config, PipelineContext(run_id="test"))
```
Comment (LucasRoesler) [inline datahub_sap_hana/ingestion.py:298]: I like this approach for have a way to manually run the code an print stuff

File: schema.py
Comment (LucasRoesler) [inline schema.py:1]: I think we need to consider renaming the file and maybe one or two of the types. This is the lineage_types or lineage_schema and i think that some of the types, like Entity either aren't needed or should be called what they are for this very specific situation: database, tables, and columns

**Review by LucasRoesler — CHANGES_REQUESTED** [review]
I think the most important thing you can do first is to add doc strings to all of your functions.

File: datahub_sap_hana/ingestion.py [hunk 11]
```diff
@@ -178,3 +203,178 @@ def _get_view_lineage_workunits(self) -> Iterable[MetadataWorkUnit]:
...
+                    view_sql = inspector.get_view_definition(view_name, schema_name)
+
+                    if view_sql:
+                        yield (view_name, view_sql)
+
+    def _get_column_lineage_for_view(
+        self, view_name: str, view_sql: str
+    ) -> Tuple[str, List[Node]]:
+        """Extracts the columns and the sql definitions of a downstream view.
+        These are then passed to the lineage function create a lineage graph of the column of an sql query.
```
Comment (LucasRoesler) [inline datahub_sap_hana/ingestion.py:225]: it is worth pointing out that the lineage function comes from a package
```suggestion
        These are then passed to the sqlglot lineage function create a lineage graph of the column of an sql query.
```

File: datahub_sap_hana/ingestion.py [hunk 12]
```diff
@@ -147,15 +173,14 @@ def _get_view_lineage_elements(self) -> Dict[Tuple[str, str], List[str]]:
...
-
         return lineage_elements
 
     def _get_view_lineage_workunits(self) -> Iterable[MetadataWorkUnit]:
         """Creates MetadataWorkUnit objects for table lineage based on the downstream and downstream objects from the query results.
         Returns an iterable MetadataWorkUnit that are emitted to Datahub.
         """
-
-        lineage_elements = self._get_view_lineage_elements()
+        conn = self.get_db_connection()
```
Comment (LucasRoesler) [inline datahub_sap_hana/ingestion.py:182]: I would actually do this one level higher, in `get_workunits` and then pass the `conn` here

File: datahub_sap_hana/ingestion.py [hunk 13]
```diff
@@ -178,3 +203,178 @@ def _get_view_lineage_workunits(self) -> Iterable[MetadataWorkUnit]:
...
+        lineages = [
+            lineage(col_name.lower(), view_sql) for col_name in selected_columns
+        ]
+        return view_name, lineages
+
+    def parse_column_name(self, column_name: str):
+        """Removes the table/name and alias from the column name returned by sqlglot, eg: instead of 'hotel.room', it will return 'room'."""
+
+        parsed_name = column_name.split(".")
+        return parsed_name[1]
```
Comment (LucasRoesler) [inline datahub_sap_hana/ingestion.py:241]: there is a potential bug here, the string might not contain a `.`, in that situation you `parsed_name = ['field_name']`, it only contains a single value, so `parsed_name[1]` will produce an error. You can use `parsed_name[-1]` which always works and means "the last element"

File: datahub_sap_hana/ingestion.py [hunk 14]
```diff
@@ -178,3 +203,178 @@ def _get_view_lineage_workunits(self) -> Iterable[MetadataWorkUnit]:
...
+            )
+            for lineage_node in column_lineages:
+                downstream_fields = Field(name=lineage_node.name)
+
+                upstream_fields_list: List[UpstreamField] = []
+
+                for column in lineage_node.downstream:
+                    upstream_fields = UpstreamField(
+                        name=self.parse_column_name(column.name),
+                        entity=Entity(name=column.source.name, platform="hana"),
```
Comment (LucasRoesler) [inline datahub_sap_hana/ingestion.py:265]: I think the platform value can come from the config maybe? there isn't an issue here _but_ it makes the code easier to copy and paste and reuse in another provider.

File: datahub_sap_hana/ingestion.py [hunk 15]
```diff
@@ -178,3 +203,178 @@ def _get_view_lineage_workunits(self) -> Iterable[MetadataWorkUnit]:
...
+
+        upstream_type = FineGrainedLineageUpstreamType.FIELD_SET
+        downstream_type = FineGrainedLineageDownstreamType.FIELD
+
+        for (
+            downstream_field,
+            upstream_fields,
+        ) in self.get_column_view_lineage_elements():
+            downstream_dataset_urn = builder.make_dataset_urn(
+                platform="hana", name=downstream_field.name, env="PROD"
```
Comment (LucasRoesler) [inline datahub_sap_hana/ingestion.py:291]: `env` definitely needs to come from the config, not a constant string.

File: schema.py [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+        List of column lineages.
+    entity : Entity
+        Entity of the lineage, which is typically a dataset.
+    """
+
+    column_lineage: List[ColumnLineage]
+    entity: Entity
+
+
+class Lineage(BaseModel):
```
Comment (LucasRoesler) [inline schema.py:88]: I think you should remove the models that you are not using, I think that include the Lineage and LineageNode, but you should double check.

File: pyproject.toml [hunk 17]
```diff
@@ -50,4 +54,8 @@ typeCheckingMode = "basic"
 
 [tool.ruff]
 # See https://beta.ruff.rs/docs/configuration/
-select = ["E", "F", "I001"]
\ No newline at end of file
+select = ["E", "F", "I001"]
+
+[tool.unimport]
+sources = ["datahub_sap_hana", "tests"]
+exclude = '__init__.py|.venv/'
```
Comment (LucasRoesler) [inline pyproject.toml:61]: oops, i forgot a new-line here, can you add one (just FYI that is what this red symbol means, a file ending without a trailing new-line)

**Review by LucasRoesler — CHANGES_REQUESTED** [review]

File: datahub_sap_hana/hana_recipe.yaml
Comment (eyelesbarrow) [reply datahub_sap_hana/hana_recipe.yaml:1]: changed and renamed the recipes: 1 that outputs to datahub, one for console

File: schema.py
Comment (eyelesbarrow) [reply schema.py:1]: I removed the classes that were not used. Only Field, Upstream and UpstreamField remain

File: datahub_sap_hana/ingestion.py [hunk 18]
```diff
@@ -178,3 +205,182 @@ def _get_view_lineage_workunits(self) -> Iterable[MetadataWorkUnit]:
...
+            for lineage_node in column_lineages:
+                downstream_fields = Field(name=lineage_node.name)
+
+                upstream_fields_list: List[UpstreamField] = []
+
+                for column in lineage_node.downstream:
+                    upstream_fields = UpstreamField(
+                        name=self.parse_column_name(column.name),
+                        column=Column(name=column.source.name,
+                                      platform=self.platform),
```
Comment (LucasRoesler) [inline datahub_sap_hana/ingestion.py:271]: I was just looking at the docs and the `@platform_name` on [L102](https://github.com/contiamo/datahub-sap-hana/pull/46/files#diff-5876beb70b85912b2f979e0f4a646e791b621b78345936702e362b58e2ed9d20R102) actually gives us a couple of methods 

```
self.get_platform_name()
self.get_platform_id()
```

I think we should 

1. Change L102 to use `@platform_name("SAP Hana")`
2. use `self.get_platform_id()` here 
```suggestion
                                      platform=self.get_platform_id()),
```

File: datahub_sap_hana/ingestion.py [hunk 19]
```diff
@@ -178,3 +205,182 @@ def _get_view_lineage_workunits(self) -> Iterable[MetadataWorkUnit]:
...
+
+        upstream_type = FineGrainedLineageUpstreamType.FIELD_SET
+        downstream_type = FineGrainedLineageDownstreamType.FIELD
+
+        for (
+            downstream_field,
+            upstream_fields,
+        ) in self.get_column_view_lineage_elements(conn):
+            downstream_dataset_urn = builder.make_dataset_urn(
+                platform=self.platform, name=downstream_field.name, env=self.config.env
```
Comment (LucasRoesler) [inline datahub_sap_hana/ingestion.py:297]:
```suggestion
                platform=self.get_platform_id(), name=downstream_field.name, env=self.config.env
```

File: datahub_sap_hana/ingestion.py [hunk 20]
```diff
@@ -178,3 +205,182 @@ def _get_view_lineage_workunits(self) -> Iterable[MetadataWorkUnit]:
...
+            )
+
+            for upstream_field in upstream_fields:
+                upstream_dataset_urn = builder.make_dataset_urn(
+                    platform=self.platform, name=upstream_field.name, env=self.config.env)
```
Comment (LucasRoesler) [inline datahub_sap_hana/ingestion.py:302]:
```suggestion
                    platform=self.get_platform_id(), name=upstream_field.name, env=self.config.env)
```

File: datahub_sap_hana/ingestion.py (hunk 18 above)
Comment (LucasRoesler) [reply datahub_sap_hana/ingestion.py:271]: It is also worth noting, if you still want the platform id to be `hana` and not `sap-hana`, which is what it will generation, then you can override the id field

```python
def platform_name(
    platform_name: str, id: Optional[str] = None, doc_order: Optional[int] = None
)
```


meaning, you can use 

```python
@platform_name(platform_name="SAP Hana", id="hana")
```

File: LICENSE
Comment (LucasRoesler) [inline LICENSE:1]: @eyelesbarrow why did you change this file?

File: datahub_sap_hana/ingestion.py [hunk 21]
```diff
@@ -63,10 +88,15 @@ class HanaConfig(BasicSQLAlchemyConfig):
     """Represents the attributes needed to configure the SAP HANA DB connection"""
 
     scheme = "hana"
-    schema_pattern: AllowDenyPattern = Field(default=AllowDenyPattern(deny=["*SYS*"]))
-    include_view_lineage: bool = Field(
+    schema_pattern: AllowDenyPattern = PydanticField(
+        default=AllowDenyPattern(deny=["*SYS*"])
+    )
+    include_view_lineage: bool = PydanticField(
```
Comment (LucasRoesler) [inline datahub_sap_hana/ingestion.py:94]: There is a merge conflict with main here, can you update your branch? Also make sure to run `poetry lock` because there is also a conflict in the lock file.

File: tests/integration/test_lineage.py [hunk 22]
```diff
@@ -0,0 +1,180 @@
...
+    upstream_field_names = [
+        "name",
+        "city",
+        "type",
+        "price"]
+    upstreams = [x[1][0].name for x in column_lineage]
+    assert upstreams == upstream_field_names, f"{upstreams}"
+
+
+"""
```
Comment (LucasRoesler) [inline tests/integration/test_lineage.py:120]: I think we should sync and cleanup the test file before we merge

File: sap_hana_results.json
Comment (LucasRoesler) [inline sap_hana_results.json:1]: this file should be removed and then we should add it to the gitignore

File: console_recipe.yaml [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+      include_field_histogram: true
+      include_field_sample_values: true
+    domain:
+      "urn:li:domain:sales":
+        allow:
+          - "HOTEL"
+sink:
+  type: file
+  config: 
+    filename: ./sap_hana_results.json
```
Comment (LucasRoesler) [inline console_recipe.yaml:36]: We should swap back to the console sink

File: tests/test_helpers/mce_helpers.py [hunk 24]
```diff
@@ -1,9 +1,3 @@
-# This is copied and slightly modified from the acryl-datahub project
-# Copyright 2015 LinkedIn Corp. All rights reserved.
-# The original source code can be found at https://github.com/datahub-project/datahub/blob/master/metadata-ingestion/tests/test_helpers/mce_helpers.py
-# And the original Apache 2 license is available at https://github.com/acryldata/datahub/blob/master/LICENSE
-
```
Comment (LucasRoesler) [inline tests/test_helpers/mce_helpers.py:5]: This shouldn't be removed

File: tests/docker-compose.yml
Comment (LucasRoesler) [inline tests/docker-compose.yml:1]: not sure how this happend, let's revert these changes to the license header

**Review by LucasRoesler — APPROVED** [review]
